### PR TITLE
Refactor popularity SQL

### LINKED ISSUE #2678: Create a mapping of media type to SQL constants
## Current Situation

See comments [here](https://github.com/WordPress/openverse/pull/2592#discussion_r1263840492) and [here](https://github.com/WordPress/openverse/pull/2592#discussion_r1264269559).

In many of our SQL tasks we have code that looks like this:

```
# Default variables to correspond to the `image` media type
def create_standardized_media_popularity_function(
    postgres_conn_id,
    media_type=IMAGE,
    function_name=STANDARDIZED_IMAGE_POPULARITY_FUNCTION,
    popularity_constants=IMAGE_POPULARITY_CONSTANTS_VIEW,
):
    # In the function body, hard-coded update if the media type differs
    if media_type == AUDIO:
        popularity_constants = AUDIO_POPULARITY_CONSTANTS_VIEW
        function_name = STANDARDIZED_AUDIO_POPULARITY_FUNCTION
```

## Suggested Improvement

We should define a mapping of media types to each of the constants needed in SQL operations (table and view names, function names, columns, etc). We should then get the values for these constants from the dictionary. Tests should also be updated.

## Benefit

Code clean-up, but this was also make it easier to add additional media types in the future.

### PR DESCRIPTION
**Pull request by stacimc**
## Fixes

Fixes #2678 by @stacimc 

## Description

This PR is entirely clean-up/refactor and should have no functional changes to the catalog.

Several large files were moved, and some appear as entirely new files instead of as 'renamed' in the diff, contributing to a misleadingly large line number count. There are many changes just updating method signatures in tests, which also add a lot of lines. The review guide below goes through each file and explains the changes to make this easier to spot.

At a high level, the actual changes made are:
* I created decorators that are used to fill in media-type-specific kwargs in SQL tasks. This replaces old, duplicated code in every one of these functions that would check the media type and set these params.
* I updated the loader and popularity SQL to use these decorators
* I removed most of the functionality of the `recreate_<media_type>_popularity_calculation` DAGs. These now *just* drop the popularity functions and recreate them. This allows the DAG to be used for updating any code changes made to the percentile or standardized popularity functions, but it won't affect the metrics table or constants.
* I dropped all references to the audio and image matviews, that are no longer being used
* Very minor cleanup like adding types and converting tasks to use TaskFlow

## Testing Instructions

Check out this branch and try `just recreate`, to thoroughly test building the schema and loading sample data.

Now try running:
* At least one audio and one image provider DAG. Suggestions: jamendo, SMK. You can mark the `pull_data` as successful as soon as a few 100 lines are written; we mostly want to test the loader steps.
* Both `recreate_<media>_popularity_calculation` DAGs. 
* Both data refreshes. Remember to enable the `create_<media_type>_filtered_index` DAGs before starting the data refreshes.
* Both popularity refreshes. Remember to enable the `batched_update` DAG before starting these.

## Code review guide

To review the code, here's an explanation of what changed file by file.

The only real new code is in creating the decorators. Recommended reviewing order:

* dags/common/constants.py: _creates a new SQLInfo dataclass for storing media-type-specific SQL information (table names, function names). Sets these up for audio and image._
* dags/common/utils.py: _Creates a decorator factory for creating decorators that will supply media-type-specific params to a function. Example usage is given in docs, and the factory is also used in this file to create the main decorator for supplying media-type-specific SQLInfo_
* dags/common/storage/db_columns.py: _Create a decorator for supplying media-type-specific DB columns to functions_
* dags/common/storage/tsv_columns.py: _Create a decorator for supplying media-type-specific TSV columns to functions_
* dags/popularity/sql.py: _This is not new code, it was just moved from dags/common/popularity/sql.py. The tasks were all updated to use TaskFlow decorators, and to get media-type specific SQLInfo from the decorators. I also dropped some now unused code (related to creating the matview which no longer exists)._
* dags/common/loader/sql.py: _Updates functions to get media-type-specific params from decorators_

Other, minor updates (mostly moving stuff around):

* DAGS.md: _updates the documentation for the ‘recreate_popularity_calculation` DAGs to note that they no longer drop any views or refresh standardized scores._
* dags/common/popularity/*: _Files were deleted or moved to dags/popularity, as they are no longer shared by any other DAGs_
* dags/common/sql.py: _Just removing an _ prefix from a utility method_
* dags/data_refresh/dag_factory.py: _Just updates use of renamed utility_
* dags/data_refresh/recreate_popularity_calculation_dag_factory.py: _This got moved to dags/popularity_
* dags/popularity/popularity_refresh_dag_factory.py: _Some code got pulled out into sql.py. The popularity refresh steps were also moved into this file for simplicity, since they are no longer shared._
* dags/popularity/popularity_refresh_types.py: _Previously, the popularity metrics dicts were hard-coded in popularity/sql.py. I moved these to be configuration that’s part of the PopularityRefresh, so all popularity configuration is in one (hopefully intuitive) place._
* dags/popularity/recreate_popularity_calculation_dag_factory.py: _This file was moved from dags/data_refresh. In addition, I took out all the tasks related to dropping and recreating the constants view (which no longer exists) and the metric table (which we *never* want to drop). So this DAG now just drops and recreates the SQL functions._
* dags/popularity/refresh_popularity_metrics_task_factory.py: _This was previously shared code. It was deleted and the popularity tasks are now part of the popularity dag factory._
* docker/upstream_db/*: _Remove materialized views_
* load_sample_data.sh: _Do not try to refresh the now dropped view._

Test stuff:

* tests/dags/common/conftest.py: _All these deleted fixtures were simply moved up a level to dags/conftest.py so they are accessible in more tests_
* tests/dags/common/loader/test_sql.py: _Just updating method signatures in tests_
* tests/dags/common/popularity/*: _Moved to tests/dags/popularity_
* tests/dags/conftest.py: _Test fixtures were moved to this file_
* tests/dags/popularity/test_popularity_refresh_types.py: _An existing test was pulled out into this file_
* tests/dags/popularity/test_sql.py: _Mostly, updating the test data setups and calling the decorated functions correctly. Tests related to the matview were dropped._
* tests/dags/test_dag_parsing: _Add init tests for popularity dags_
* tests/test_utils/sql.py: _Update constants_

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: catalog/dags/popularity/popularity_refresh_dag_factory.py [hunk 1]
```diff
@@ -70,52 +67,6 @@ def get_last_updated_time():
     return datetime.utcnow()
 
 
-@task
```
Comment (stacimc) [inline catalog/dags/popularity/popularity_refresh_dag_factory.py:73]: Review note: This task was moved into `sql.py`, not deleted.

**Comment by sarayourfriend** [comment]
@stacimc Just letting you know that I've seen this PR and will review it tomorrow. My brain is fried a the moment and I won't be able to comprehend this right but, but thanks in advance for what look like an excellent description and test instructions. Your PR descriptions are second to none and I deeply appreciate them :pray:

File: catalog/dags/common/constants.py [hunk 2]
```diff
@@ -36,3 +37,41 @@
...
+        standardized_popularity_fn="standardized_audio_popularity",
+        popularity_percentile_fn="audio_popularity_percentile",
+    ),
+    IMAGE: SQLInfo(
+        media_table=IMAGE,
+        metrics_table="image_popularity_metrics",
+        standardized_popularity_fn="standardized_image_popularity",
+        popularity_percentile_fn="image_popularity_percentile",
+    ),
+}
```
Comment (sarayourfriend) [inline catalog/dags/common/constants.py:77]: Awesome. There's a similar dataclass in the API's testconf: https://github.com/WordPress/openverse/blob/HEAD/api/test/unit/conftest.py#L59-L92

File: catalog/dags/common/utils.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from common.constants import SQL_INFO_BY_MEDIA_TYPE
+
+
+def setup_kwargs_for_media_type(
+    values_by_media_type: dict[str, Any], kwarg_name: str
+) -> callable:
+    """
+    Create a decorator which provides media_type-specific information as parameters
+    for the called function. The called function must itself have a media_type kwarg,
+    which is used to select values.
```
Comment (sarayourfriend) [inline catalog/dags/common/utils.py:13]: Not sure if it's worth it, but thought I'd mention that `inspect.signature` can list the parameters for a function, including explicitly defined kwargs. If desirable, a configuration check could be added in `wrap` that confirms the function is correctly configured for the decorator (including that `kwarg_name` is present). This wouldn't cover `**kwargs` use cases though.

File: catalog/dags/common/utils.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import functools
+from typing import Any
+
+from common.constants import SQL_INFO_BY_MEDIA_TYPE
+
+
+def setup_kwargs_for_media_type(
+    values_by_media_type: dict[str, Any], kwarg_name: str
+) -> callable:
```
Comment (sarayourfriend) [inline catalog/dags/common/utils.py:9]: Question regarding the implementation: is the customisable `kwarg_name` a byproduct of existing implementations? I'm wondering specifically why we don't always use the `media_info` kwarg name. Is it because existing implementations use a different name and you wanted to avoid making changes to those names?

Edit: Never mind. I misunderstood how this function was used, but after reading `setup_sql_info_for_media_type` at the end, I get now why `kwarg_name` is configurable. It's dependent on what the "values" are in `values_by_media_type` :+1: Disregard this comment. Leaving it intact just in case someone else runs into the same misconception about how this works.

File: catalog/dags/common/storage/db_columns.py [hunk 5]
```diff
@@ -79,3 +81,10 @@
     col.STANDARDIZED_POPULARITY,
     col.AUDIO_SET_FOREIGN_IDENTIFIER,
 ]
+
+DB_COLUMNS_BY_MEDIA_TYPE = {AUDIO: AUDIO_TABLE_COLUMNS, IMAGE: IMAGE_TABLE_COLUMNS}
+
+
+def setup_db_columns_for_media_type(func: callable) -> callable:
+    """Provide media-type-specific DB columns as a kwarg to the decorated function."""
+    return setup_kwargs_for_media_type(DB_COLUMNS_BY_MEDIA_TYPE, "db_columns")(func)
```
Comment (sarayourfriend) [inline catalog/dags/common/storage/db_columns.py:90]: FWIW, these implementations can be simplified a bit to:

```suggestion
setup_db_columns_for_media_type = setup_kwargs_for_media_type(DB_COLUMNS_BY_MEDIA_TYPE, "db_columns")
"""Provide media-type-specific DB columns as a kwarg to the decorated function."""
```

I'm not sure if that results in a worse docstring experience in the editor, though.

File: catalog/dags/popularity/sql.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+@task
+@setup_sql_info_for_media_type
+def update_percentile_and_constants_values_for_provider(
+    postgres_conn_id: str,
+    provider: str,
+    raw_percentile_value: float,
+    media_type: str,
+    popularity_metrics: dict,
+    sql_info: SQLInfo = None,
```
Comment (sarayourfriend) [inline catalog/dags/popularity/sql.py:158]: Is the `None` default required here? Won't it always be supplied by `@setup_sql_info_for_media_type`? Is the default to satisfy a typecheck for the caller, I guess so editors don't think the callsite _needs_ to pass that kwarg?

I guess if the decorator implemented [ParamSpec](https://docs.python.org/3/library/typing.html#typing.ParamSpec) on the `callable` type of the wrapped function and itself set the kwarg to optional it could handle the call site documentation. Complex though, so not worth it. It does a bit of confusion in the function implementation though, as `sql_info` would never _actually_ be `None`... the decorator itself overrides the value, even if the caller explicitly set it to `None`. I'd be tempted to check that `sql_info is not None` if I was working in this function and didn't have a nuanced understanding of how `setup_sql_info_for_media_type` worked.

To summarise: the `None` default is confusing when making changes to the function, if the decorator's implementation details and implications1 aren't well understood. The risk of error is low (checking `sql_info is not None` won't hurt, it's just technically useless) so it isn't a blocking issue, just a nit pick. Considering the complexity that it could take to solve this, if editors complain about not passing `sql_info`, it might not be worth it to make any changes here. On the other hand, if the default parameter isn't necessary to satisfy basic editor introspection, it would improve the clarity of these functions to remove the default `None` value.

File: catalog/dags/popularity/sql.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+        """
+    )
+    postgres.run(query)
+
+
+@setup_sql_info_for_media_type
+def format_update_standardized_popularity_query(
+    media_type: str,
+    sql_info: SQLInfo = None,
+    task: AbstractOperator = None,
```
Comment (sarayourfriend) [inline catalog/dags/popularity/sql.py:309]: Why is `task` passed?

**Review by sarayourfriend — COMMENTED** [review]
That new decorator is a nifty abstraction! It's clear how much it cleans up the code that relies on it. Nice work identifying that.

Just leaving comments before testing locally, but doing that local testing right now.

**Review by sarayourfriend — APPROVED** [review]
LGTM. Like I said, great clean up, nice work identifying opportunities for clean and sensible abstractions :rocket:

File: catalog/dags/common/utils.py (hunk 3 above)
Comment (stacimc) [reply catalog/dags/common/utils.py:13]: This is great! I struggled a bit with this, because `media_type` doesn't just need to be a keyword argument in the signature -- it needs to be passed to the called function as a keyword argument, never as a positional one. Calling this out has made me realize I should update it to be a [keyword-only argument](https://peps.python.org/pep-3102/#:~:text=Keyword%2Donly%20arguments%20are%20not,therefore%20'required%20keyword'%20arguments.), and it looks like I should be able to add a configuration check for that explicitly :)

File: catalog/dags/popularity/sql.py (hunk 6 above)
Comment (stacimc) [reply catalog/dags/popularity/sql.py:158]: I agree on all counts -- unfortunately, the `None` default does appear to be required when the function is also using the `@task` decorator for TaskFlow, or we get errors at DAG parsing :(

File: catalog/dags/popularity/sql.py (hunk 7 above)
Comment (stacimc) [reply catalog/dags/popularity/sql.py]: No need! Removed, good catch.

File: catalog/dags/common/storage/db_columns.py (hunk 5 above)
Comment (stacimc) [reply catalog/dags/common/storage/db_columns.py:90]: Totally! I didn't like the way the docstrings looked, but I actually have no idea what the "ideal" solution is here, that's obviously personal preference. I do think these merit the docstring either way, though.

File: catalog/dags/common/utils.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        use that value instead.
+        """
+
+        # The called function must be supplied a `media_type` keyword-only argument. It
+        # cannot allow the value to be supplied as a positional argument.
+        if (
+            media_type := signature(func).parameters.get("media_type")
+        ) is None or media_type.kind != _ParameterKind.KEYWORD_ONLY:
+            raise Exception(
+                "Improperly configured: `media_type` must be a keyword-only argument"
```
Comment (obulat) [inline catalog/dags/common/utils.py:47]: Would it be useful for debugging down the road to add the name of what is _improperly configured_ here? It's the wrapped function, right?

File: catalog/dags/common/utils.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    def wrap(func: callable) -> callable:
+        """
+        Provide the appropriate value for the media_type passed in the called function.
+        If the called function is already explicitly passed a value for `kwarg_name`,
+        use that value instead.
+        """
+
+        # The called function must be supplied a `media_type` keyword-only argument. It
+        # cannot allow the value to be supplied as a positional argument.
+        if (
```
[the remark this answers is not in the document]
Comment (stacimc) [reply catalog/dags/common/utils.py:43]: This isn't checking the value of the `media_type` -- this check just validates the method signature of the wrapped function to check that `media_type` is a keyword-only argument.

File: catalog/dags/common/utils.py (hunk 8 above)
Comment (stacimc) [reply catalog/dags/common/utils.py:47]: Yes, that's a great point! Adding this in.
